fix(ivy): ngcc should not emit TypeScript syntax

JoostK opened the pull request:
Previously, ngcc used ngtsc's logic to transform the Angular AST
to a TypeScript AST and then leverages TypeScript's node printer to
obtain the source text to insert. This turns out to be a problem when
specific TypeScript syntax is represented in the AST; the printer cannot
be configured to output only JavaScript syntax.

A full-blown TypeScript emit of a complete `ts.SourceFile` would be
required to be able to emit JS and possibly downlevel into a lower
language target, which is not an option for ngcc as it currently
operates on partial ASTs, not full source files.

Instead, the Angular AST itself is used to emit source files, using the
emitter logic that is also used in JIT compilations. As the Angular AST
produced by ngtsc may contain literal TypeScript nodes inside of
`WrappedNodeExpr`, a TypeScript printer is still used to obtain source
text for such nodes (assuming no TypeScript specific syntax is present
within that node's subtree).

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

ngcc may output TypeScript specific syntax into JavaScript bundles, which fail to be parsed.

## What is the new behavior?

ngcc uses existing emitter logic that operates on Angular AST to produce JavaScript code instead of TypeScript.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

@alxhub Can you comment on my assumption that we can safely print `WrappedNodeExpr` using TypeScript's printing logic, as doing so still has a risk of producing TS specific syntax.

petebacondarwin commented:
> @alxhub Can you comment on my assumption that we can safely print WrappedNodeExpr using TypeScript's printing logic, as doing so still has a risk of producing TS specific syntax.

I am pretty confident that for ngcc, all of these wrapped nodes will have come from JS files, so it seems likely that they are already JS. Right?

petebacondarwin commented on packages/compiler/src/output/abstract_js_emitter.ts:
```diff
@@ -100,6 +100,9 @@ export abstract class AbstractJsEmitterVisitor extends AbstractEmitterVisitor {
   visitInvokeFunctionExpr(expr: o.InvokeFunctionExpr, ctx: EmitterVisitorContext): string|null {
     const fnExpr = expr.fn;
+    if (expr.pure) {
+      ctx.print(expr, '/*@__PURE__*/');
```
It seems that there is test that is breaking due to this?

https://travis-ci.org/angular/angular/jobs/453577853#L1288

JoostK replied:
Oops, I added that test with a type ;)

petebacondarwin commented:
Generally looks like a good idea!

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:
```diff
@@ -345,3 +337,43 @@ function createAssignmentStatement(
+  visitExternalExpr(ast: ExternalExpr, ctx: EmitterVisitorContext): any {
+    const {name, moduleName} = ast.value;
+    if (moduleName) {
+      const prefix = this.importManager.generateNamedImport(moduleName, name !);
```
How do we know what `name` is defined? Should we not check `name` for the whole function body?

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts:
```diff
@@ -127,11 +152,14 @@ describe('Renderer', () => {
-A.ngDirectiveDef = ɵngcc0.ɵdefineDirective({ type: A, selectors: [["", "a", ""]], factory: function A_Factory(t) { return new (t || A)(); } });`);
+    }],
+    null,{ foo: [] });
+A.ngDirectiveDef = ɵngcc0.ɵdefineDirective({type:A,selectors:[['','a','']],factory:function A_Factory(t) {
```
Was the previous printer producing a fat arrow here or something?

JoostK replied:
Nope, it was only the non-null assertion in the template (`person!.name`) that tripped stuff up. It's actually surprising that the TypeScript printer has worked so well until that tiny syntactical difference.

JoostK commented:
@petebacondarwin @alxhub After a rebase this is now green again, should be good to go. Let me know if we do want to elide the `!` in ngtsc directly, if so I'll update the PR accordingly.

JoostK committed:
fix(ivy): ngcc should not emit TypeScript syntax

If a template contains specific TypeScript syntax, such as a non-null
assertion, the code that is emitted from ngcc into a JavaScript bundle
should not retain such syntax as it is invalid in JS.

A full-blown TypeScript emit of a complete ts.SourceFile would be
required to be able to emit JS and possibly downlevel into a lower
language target, which is not an option for ngcc as it currently
operates on partial ASTs, not full source files.

Instead, ngtsc no longer produces TypeScript specific syntax in the first
place, such that TypeScript print logic will only generate JS code.

petebacondarwin reviewed (dismissed):
This fix looks good to me. It needs to have an ngtsc test too IMO. I am going to add one.
Then leave it to @alxhub to do final approval.

petebacondarwin reviewed (approved):
LGTM! @alxhub over to you. I have rebased on master, just to be sure!
